src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,305 @@
+    change_targets = [A1, A2, X_gate, B1, B2, C]
+    for i in range(len(change_targets)):
+        change_targets[i].targets = [1]
+
+    # form full gate list
+    final_gate_list = [
+        A1,
+        A2,
+        X_gate,
+        CNOT_gate,
+        X_gate,
+        B1,
+        B2,
+        X_gate,
+        CNOT_gate,
+        X_gate,
+        C,
+        phase_gate]
+    return(final_gate_list)
+
+
+def _two_qubit_firstq_target(last_two_level_array_info):
+    """ Finds equivalent circuit of two qubit two-level array when last qubit
+    is the target and first is the control with a control value = 1.
+
+    This object is the first array in output from
+    _decompose_to_two_level_arrays(input_gate, num_qubits, expand = False)
+
+    Only works for ind1 = 1, ind2 = 3
+    """
+    last_two_level_qobj = last_two_level_array_info[-1]
+    last_two_level_array = last_two_level_qobj.full()
+
+    # check two_level_gate
+    check_gate((Qobj(last_two_level_array, dims=[[2] * 2] * 2)), 2)
+
+    # non-trivial indices
+    index_list = last_two_level_array_info[0]
+    index1, index2 = index_list
+    gray_code_dict = _gray_code_gate_info(index1, index2, 2)
+
+    # pauli x gates to make all control 1
+    new_gray_code_dict = _paulix_for_control_0(gray_code_dict)
+    control_qubit = new_gray_code_dict[0]['controls ='][0]
+    target_qubit = new_gray_code_dict[0]['targets ='][0]
+
+    # get ABC decomposition gates
+    single_qubit_decomposition = decompose_one_qubit_gate(
+        last_two_level_qobj, 'ZYZ_PauliX')
+    CNOT_gate = Gate("CNOT", controls=control_qubit, targets=target_qubit)
+
+    # A1, A2 (with default target 0) forming A are at indices 0, 1
+    A1 = single_qubit_decomposition[0]
+    A2 = single_qubit_decomposition[1]
+
+    # At index 2, pauli x gate with default target 0
+    X_gate = single_qubit_decomposition[2]
+
+    # B1, B2 (with default target 0) forming B are at indices 3, 4
+    B1 = single_qubit_decomposition[3]
+    B2 = single_qubit_decomposition[4]
+
+    # C (with default target 0) is at index 6
+    C = single_qubit_decomposition[6]
+
+    # last is the global phase gate which is changed to phasegate
+    phase_angle = single_qubit_decomposition[-1].arg_value
+    phase_gate = Gate(
+        "PHASEGATE",
+        targets=[1],
+        arg_value=phase_angle,
+        arg_label=r"{:0.2f} \times \pi".format(phase_angle / np.pi),
+    )
+
+    # change default targets for gates A, B, C
+    change_targets = [A1, A2, X_gate, B1, B2, C]
+    for i in range(len(change_targets)):
+        change_targets[i].targets = [0]
+
+    # form full gate list
+    final_gate_list = [
+        A1,
+        A2,
+        X_gate,
+        CNOT_gate,
+        X_gate,
+        B1,
+        B2,
+        X_gate,
+        CNOT_gate,
+        X_gate,
+        C,
+        phase_gate]
+    return(final_gate_list)
+
+
+def _find_index_for_firstq_target(two_level_array_output, num_qubits):
```

Comment:
This is for the case when the two-level array does not have first or last qubit as target.

Reply:
Re-defining the function and deleting two separate functions for first qubit and last qubit